src/docfx/lib/schema/JsonSchemaValidator.cs

```diff
@@ -18,45 +18,54 @@ internal class JsonSchemaValidator
         private readonly JsonSchema _schema;
         private readonly JsonSchemaDefinition _definitions;
         private readonly MicrosoftGraphAccessor? _microsoftGraphAccessor;
-        private readonly ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source, bool? isCanonicalVersion)> _metadataBuilder;
-        private static readonly ThreadLocal<bool?> t_isCanonicalVersion = new ThreadLocal<bool?>();
+        private readonly JsonSchemaValidatorExtension? _ext;
+        private readonly ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source, FilePath? filePath)> _metadataBuilder;
+        private static readonly ThreadLocal<FilePath?> t_filePath = new ThreadLocal<FilePath?>();
 
         public JsonSchema Schema => _schema;
 
-        public JsonSchemaValidator(JsonSchema schema, MicrosoftGraphAccessor? microsoftGraphAccessor = null, bool forceError = false)
+        public JsonSchemaValidator(
+            JsonSchema schema,
+            MicrosoftGraphAccessor? microsoftGraphAccessor = null,
+            bool forceError = false,
+            JsonSchemaValidatorExtension? ext = null)
         {
             _schema = schema;
             _forceError = forceError;
             _definitions = new JsonSchemaDefinition(schema);
             _microsoftGraphAccessor = microsoftGraphAccessor;
-            _metadataBuilder = new ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source, bool? isCanonicalVersion)>();
+            _ext = ext;
+            _metadataBuilder = new ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source, FilePath? filePath)>();
         }
 
-        public List<Error> Validate(JToken token, bool? isCanonicalVersion = null)
+        public List<Error> Validate(JToken token, FilePath? filePath = null)
```

Comment:
Make `FilePath` mandatory, there should always be a file otherwise it indicates a bug. (Tests can fake a path).
```suggestion
        public List<Error> Validate(JToken token, FilePath filePath)
```

Reply:
done